R/sampler-parallel-tempering.R

```diff
@@ -39,211 +39,328 @@
+    array_bind(arrays = lapply(state, "[[", name), ...)
   }
+  accept_swap <- join("accept_swap", on = 2)
+  even_step <- vlapply(state, "[[", "even_step")
+  hot <- lapply(state, "[[", "hot")
+  sub_state_chain <- lapply(state, "[[", "sub_state_chain")
+  sub_state_sampler <- control$sampler$state$combine(
+    lapply(state, "[[", "sub_state_sampler"))
+  list(accept_swap = accept_swap,
+       even_step = even_step,
+       hot = hot,
+       sub_state_chain = sub_state_chain,
+       sub_state_sampler = sub_state_sampler)
+}
 
-  finalise <- function(state, model, rng) {
-    list(accept_swap = internal$accept_swap)
+
+sampler_parallel_tempering_restore <- function(chain_id, state_chain,
+                                               state_sampler, control,
+                                               model) {
+  stopifnot(length(chain_id) == 1)
+
+  model_base <- parallel_tempering_base(model, control)
+  model_scaled <- parallel_tempering_scale(model, control)
+
+  sub_state_chain <- state_sampler$sub_state_chain[[chain_id]]
+  sub_state_sampler <- control$sampler$state$restore(
+    chain_id,
+    sub_state_chain,
+    state_sampler$sub_state_sampler,
+    control$sampler$control,
+    model_scaled)
+  hot <- parallel_tempering_hot(model, control, state_sampler$hot[[chain_id]])
+  even_step <- state_sampler$even_step[[chain_id]]
+  accept_swap <- state_sampler$accept_swap[, chain_id]
+
+  env <- new.env(parent = emptyenv())
+  env$model_scaled <- model_scaled
+  env$model_base <- model_base
+  env$sub_state_chain <- sub_state_chain
+  env$sub_state_sampler <- sub_state_sampler
+  env$hot <- hot
+  env$even_step <- even_step
+  env$accept_swap <- accept_swap
+  env
+}
+
+
+sampler_parallel_tempering_details <- function(state, control) {
+  list(accept_swap = state$accept_swap,
+       sampler = control$sampler$state$details(state$sub_state_sampler))
+}
+
+
+## A caching direct sampler; we draw 'size' samples from the base
+## distribution (perhaps the prior) and compute the true model density
+## for all of these at once, then return these one at a time (so the
+## actual calculation only happens every 'size' steps).  This means
+## that if the model is configured to run in parallel with 'n_rungs'
+## threads (which will be the same as 'size') then calculating from
+## the prior will be as efficient as the samples from the other
+## chains.
+parallel_tempering_hot <- function(model, control, state = NULL) {
+  model_base <- parallel_tempering_base(model, control)
+  size <- control$n_rungs
+
+  env <- new.env(parent = emptyenv())
+  env$index <- size
+
+  if (!is.null(state)) {
+    list2env(state, env)
   }
 
-  get_internal_state <- function() {
-    keep <- c("accept_swap", "even_step", "last", "state")
-    set_names(lapply(keep, function(k) internal[[k]]), keep)
+  sample <- function(rng) {
+    if (env$index >= size) {
+      env$pars <- direct_sample_many(size, model_base, rng)
+      env$density <- model$density(env$pars)
```

Comment:
if e.g. the model is of form `prior + likelihood` where likelihood is from `dust_likelihood_monty`, does `model$density(env$pars)` involve running the filter/unfilter? so potentially costly?